### ISSUE TITLE
Stack overflow when type-checking a class that inherits from a `pyside6-uic` generated UI mixin

### ISSUE THREAD

**Issue by aqc-carlodri** [astral-sh/ty#3822]
> _This issue was written by 🤖 [Claude Code](https://claude.ai/code) (claude-sonnet-4-6) under the guidance of the repository owner, who performed the debugging session that produced the bisection evidence above._

## Summary

ty 0.0.51 aborts with a fatal stack overflow when a form class inherits from both a Qt widget and a `pyside6-uic`-generated UI mixin, and a method on that class accesses an attribute that is assigned **late** in the generated `setupUi` body.

```
thread '<unknown>' (30267) has overflowed its stack
fatal runtime error: stack overflow, aborting
```

No diagnostic is emitted — the process just aborts.

---

## Environment

| | |
|---|---|
| ty | 0.0.51 |
| Python | 3.13 (pixi managed) |
| PySide6 | 6.11.1 (conda-forge) |
| Platform | linux-64 |
| Command | `ty check src tests` |

---

## Root cause

This is a **stack-depth exhaustion bug**, not a circular type reference.

When ty resolves the type of `self.target_spin` in `MyForm.foo`, it must find where that attribute is assigned. It scans `Ui_MyForm.setupUi` from the top — and every PySide6 method call it passes through (`.setObjectName()`, `.setMinimum()`, etc.) requires resolving method signatures from the PySide6 stubs, which pull in Qt's deep inheritance chains. Recursion depth grows proportionally with the number of preceding statements.

```
resolve  MyForm.foo  →  lookup self.target_spin
  └─ scan Ui_MyForm.setupUi body from top
       ├─ line    1: self._root = QVBoxLayout(MyForm)     ← ok
       ├─ line    2: self._w0   = QDoubleSpinBox(MyForm)  ← ok
       ├─ line    3: self._w0.setObjectName(...)           ← ok
       ├─ ...
       ├─ line 3000: self._w749.setMaximum(100.0)          ← stack ~90%
       ├─ ...
       └─ line 3283: self.target_spin = QSpinBox(MyForm)  ← OVERFLOW
```

The crash was confirmed via binary search on the `setupUi` body of our 3 592-line generated file:

- Truncated to **3 282 lines** → ✅ passes  
- Truncated to **3 283 lines** → 💥 crashes — line 3 283 is exactly `self.max_bin = QSpinBox(…)`, the attribute accessed by the failing method

**Two conditions are necessary:**
1. The Ui class must be a **resolvable cross-module import** (inlining the same 3 592-line class in the same file passes — the cross-module path adds extra stack depth).
2. The accessed attribute must be defined **after ~3 283 lines** of the `setupUi` body.

This affects any project that uses `pyside6-uic` to generate forms with many widgets and accesses attributes of widgets defined late in the generated file.

---

## How to reproduce

The following script generates a self-contained repro directory. No private packages required.

```python
#!/usr/bin/env python3
# generate_repro.py
# Run:  python generate_repro.py && ty check repro/consumer.py
import os

os.makedirs("repro", exist_ok=True)

# repro/__init__.py — makes 'repro' a resolvable package
open("repro/__init__.py", "w").close()

# repro/ui_generated.py
# Mimics pyside6-uic output: target attribute defined after enough
# preceding PySide6 method calls to exhaust the stack.
LINES_NEEDED = 3300  # empirical threshold; increase if your stack is larger

with open("repro/ui_generated.py", "w") as f:
    f.write("from PySide6.QtWidgets import "
            "QDoubleSpinBox, QLabel, QSpinBox, QVBoxLayout, QWidget\n\n")
    f.write("class Ui_MyForm:\n")
    f.write("    def setupUi(self, MyForm):\n")
    f.write("        self._root = QVBoxLayout(MyForm)\n")
    # Padding: enough QDoubleSpinBox assignments + method calls to fill the stack
    for i in range(LINES_NEEDED // 4):
        f.write(f"        self._w{i} = QDoubleSpinBox(MyForm)\n")
        f.write(f"        self._w{i}.setObjectName('w{i}')\n")
        f.write(f"        self._w{i}.setMinimum(0.0)\n")
        f.write(f"        self._w{i}.setMaximum(100.0)\n")
    # Target attribute defined LAST — ty must scan the whole body to find it
    f.write("        self.target_spin = QSpinBox(MyForm)\n")
    f.write("        self.target_spin.setMaximum(255)\n")
    f.write("    def retranslateUi(self, MyForm): pass\n")

# repro/consumer.py
with open("repro/consumer.py", "w") as f:
    f.write("from PySide6.QtWidgets import QWidget\n")
    f.write("from repro.ui_generated import Ui_MyForm\n\n")
    f.write("class MyForm(QWidget, Ui_MyForm):\n")
    f.write("    def foo(self) -> None:\n")
    f.write("        x = self.target_spin.value()  # triggers stack overflow\n")

print("Generated repro/. Run:  ty check repro/consumer.py")
```

> **Note on `LINES_NEEDED`:** The exact threshold depends on the platform's default stack size. On our linux-64 system the crash first occurs at `setupUi` body line 3 283. Use 3 300 as a safe margin; increase if your system has a larger stack.

---

## Bisection log

| Target / condition | Result |
|---|---|
| Full project (`src/ tests/`) | 💥 CRASH |
| `src/core/` | ✅ PASS |
| `src/data_processing/` | ✅ PASS |
| `src/ui/` | 💥 CRASH |
| Each `src/ui/*.py` individually (10 files) | 9× ✅ PASS · 1× 💥 CRASH (`lens_aperture.py`) |
| All imports in `lens_aperture.py`, no class body | ✅ PASS |
| Class with empty methods only | ✅ PASS |
| Method `_update_histogram_range` in isolation | 💥 CRASH |
| `QSpinBox().value()` on a local variable | ✅ PASS |
| `self.max_bin.value()` — without Ui base class | ✅ PASS (type error only) |
| `self.max_bin.value()` — small hand-written Ui mixin | ✅ PASS |
| `self.max_bin.value()` — Ui class inlined in same file (3 592 lines) | ✅ PASS |
| `self.max_bin.value()` — Ui class imported from package (3 592 lines) | 💥 CRASH |
| `setupUi` body truncated to 3 282 lines (package import) | ✅ PASS |
| `setupUi` body truncated to 3 283 lines (package import) | 💥 CRASH ← line 3 283 = `self.max_bin = QSpinBox(…)` |

---

## Workaround

Adding explicit class-level annotations for the affected attributes causes ty to use them directly without scanning `setupUi`:

```python
from PySide6.QtWidgets import QSpinBox, QWidget
from repro.ui_generated import Ui_MyForm

class MyForm(QWidget, Ui_MyForm):
    target_spin: QSpinBox  # explicit annotation — ty stops here, never scans setupUi

    def foo(self) -> None:
        x = self.target_spin.value()  # passes
```

This is not practical for real projects — a form with 80 widgets would need 80 annotations, duplicating information already in the generated file.

---

## Expected behavior

ty should bound the recursion depth when scanning method bodies for attribute assignments, either by imposing a depth limit and falling back to `Unknown`, or by restructuring the scan to be iterative. The process should never abort; all overflow conditions should surface as diagnostics.

---

**Comment by aqc-carlodri** [comment]
_This comment was written by [Claude Code](https://claude.ai/code) (claude-sonnet-4-6) under the guidance of the repository owner._

---

**The pyside6-uic framing in the issue title is misleading — this is a general bug.**

The `pyside6-uic` pattern is just what triggered it in our project. The underlying problem is in ty's attribute inference for any cross-module imported class: when ty needs to resolve the type of `self.x`, it scans the method bodies of the class and its bases looking for assignments to `x`. That scan is recursive, and the recursion depth grows proportionally with the number of statements it has to walk through before finding the assignment.

`pyside6-uic` generates methods that are thousands of lines long, which is why we hit it here — but the same crash would occur with any sufficiently large method that assigns instance attributes late in its body:

```python
# anything_large.py  (cross-module import required)
class BigInitMixin:
    def __init__(self):
        # ~800+ assignments, each with a few method calls...
        self.a0 = SomeType(); self.a0.configure()
        self.a1 = SomeType(); self.a1.configure()
        # ... 800 more ...
        self.target = SomeType()   # ← accessing this from a subclass crashes ty

# consumer.py
from anything_large import BigInitMixin

class MyClass(Base, BigInitMixin):
    def foo(self) -> None:
        self.target.do_something()  # stack overflow
```

The fix should be in ty's attribute assignment scanner — it needs a recursion depth guard (or an iterative implementation) that applies regardless of which library or tool generated the method body.

Worth also noting: the same class inlined in the consumer file does **not** crash, so the cross-module import path adds meaningful extra stack depth on top of the body scan itself.

🤖 Created by Claude Code

**Comment by aqc-carlodri** [comment]
_This comment was written by [Claude Code](https://claude.ai/code) (claude-sonnet-4-6) under the guidance of the repository owner._

---

This appears to be a real-world instance of #96 ("Stack overflow for large chain of inter-dependent definitions").

The mechanism described there — bindings analyzed lazily from last to first, recursively — matches exactly what we observe: ty must walk every preceding statement in `setupUi` before reaching the target attribute assignment, and each step adds a stack frame.

One additional data point our case contributes: the crash occurs at ~3 283 sequential assignments in a **cross-module import** context, whereas #96 notes the release-build threshold is ~2 500 for the inline case. The cross-module path appears to add enough extra overhead to lower the effective threshold — the same 3 592-line class inlined in the consumer file does **not** crash.

🤖 Created by Claude Code
